### ISSUE TITLE
Integrate LED²-Net as FastAPI microservice for 3D room layout extraction

### ISSUE THREAD

**Issue by Serenaweiyingwu** [Aroomyaroomy/LED2-Net#1]
## Overview

Wrap the best-performing 360° room layout estimation model in a FastAPI microservice that accepts an equirectangular panorama and returns a 3D room layout. This feeds into the Aroomy Three.js room viewer and AI Design Assistant.

No model training is required — pretrained weights are publicly available for both candidate models.

**Approach: run Phase 1 side-by-side evaluation of LED²-Net and LGT-Net on the same real-estate panoramas, then integrate whichever performs better on our actual data.**

---

### LED²-Net (CVPR 2021)
**Repo: https://github.com/Aroomyaroomy/LED2-Net** (our fork)

Reformulates layout estimation as horizon-depth prediction. The L2D transformation converts predicted boundary angles into wall planes via ray-casting, making the loss 3D-aware and the pipeline end-to-end trainable.

**L2D Transformation — 3 steps:**
1. **Layout Plane Recovering** — predicts boundary points in spherical coordinates (θ, φ), projects them onto 3D wall planes using Manhattan World assumptions (camera height = 1.6m, perpendicular walls)
2. **Grid Re-sample** — casts M = 256 rays uniformly around the 360° horizon
3. **Differentiable Depth Rendering** — intersects each ray with all wall planes, discards invalid candidates (negative depth, outside wall's angular span), takes minimum depth (closest wall wins)

### LGT-Net (CVPR 2022)
**Repo: https://github.com/Aroomyaroomy/LGT-Net** (our fork)

Successor to LED²-Net. Two key improvements:
1. **SWG-Transformer** (Shifted Window + Global attention) replaces the bi-LSTM — better at capturing both local wall transitions and global room structure simultaneously
2. **Planar-geometry loss** — adds supervision on wall normals and normal gradients, explicitly penalizing non-flat walls and soft corners. LED²-Net's depth loss only cares about distance to the wall, not whether the wall is planar

**Benchmark comparison (3D IoU):**

| Dataset | LED²-Net | LGT-Net |
|---|---|---|
| PanoContext | 82.75% | 85.49% |
| Stanford2D3D | 83.77% | 86.16% |
| Matterport3D | 81.52% | 81.76% |
| Realtor360 | 85.21% | — |

LGT-Net wins on paper, but benchmark data ≠ real-estate listing photos. Phase 1 exists to verify this on our actual data.

### Hard architectural constraints (both models)
- Walls must be flat and perpendicular (Manhattan World). L-shaped rooms, bay windows, angled walls will degrade accuracy.
- Camera height is fixed at 1.6m. Panoramas shot significantly higher/lower will have dimension errors.
- Input must be a full equirectangular panorama — regular perspective photos will not work.

---

## Phase 1 — Side-by-Side Evaluation (start here)

Run **both models on the same 10 panoramas** and compare outputs visually. Pick the winner before writing any integration code.

- [ ] Set up local environments for both LED²-Net and LGT-Net, download pretrained weights for each
- [ ] Collect 10 sample real-estate panoramas covering varied room types (bedroom, living room, kitchen, complex layout, L-shaped if possible)
- [ ] Run inference with both models on each panorama
- [ ] Visualize the 3D output for each — do the room shapes look geometrically correct?
- [ ] Note failure cases per model: which room types or shooting conditions produce bad results?
- [ ] Pick the better-performing model and post the evaluation summary (with screenshots) as a comment on this issue before proceeding to Phase 2

**Key questions to answer:**
- Does LGT-Net's benchmark advantage hold on real Aroomy listing photos, or does LED²-Net close the gap?
- Are there room types where one model clearly fails and the other doesn't?
- Is inference speed acceptable for both? (target < 5s per image on GPU)

---

## Phase 2 — Integration (pending Phase 1 results)

> **Do not start Phase 2 until the Phase 1 evaluation is reviewed and a model is chosen.**

### 2a. FastAPI microservice
- Wrap chosen model's inference in a `POST /layout` endpoint
- Containerize with Docker; requires GPU instance (e.g. `g4dn.xlarge`)

**API contract:**
```
POST /layout
Content-Type: multipart/form-data
Body: panorama (image/jpeg or image/png, equirectangular)

Response 200:
{
  "corners": [
    {"x": 2.1, "y": 0.0, "z": 1.8},
    ...
  ],
  "horizon_depth": [1.8, 2.1, 2.3, ...],  // 256 floats, meters
  "room": {
    "width": 4.2,
    "depth": 3.6,
    "height": 2.4
  },
  "num_corners": 4,
  "model": "LGT-Net"  // or "LED2-Net", whichever was chosen
}
```

### 2b. Spring Boot integration
- After panorama upload to S3, call the layout microservice
- Store returned layout JSON in RDS alongside the panorama record
- Expose via `GET /api/rooms/{id}/layout`

### 2c. Three.js viewer
- Consume corner coordinates from `/api/rooms/{id}/layout`
- Each pair of adjacent corners defines a wall — render as `PlaneGeometry`
- Floor and ceiling use `room.height`
- The 256-element `horizon_depth` array can drive a top-down 2D floor plan view

---

## Acceptance Criteria

**Phase 1**
- [ ] Both models run successfully on local GPU setup
- [ ] 10 panoramas tested through both models, results documented with screenshots
- [ ] Evaluation summary posted as a comment on this issue with a clear model recommendation

**Phase 2**
- [ ] FastAPI service runs in Docker, returns layout JSON within 5 seconds on GPU instance
- [ ] Spring Boot stores layout JSON in RDS and exposes `GET /api/rooms/{id}/layout`
- [ ] Three.js viewer renders a correct 3D room from corner coordinates for a standard 4-corner room
- [ ] Service handles graceful errors: non-equirectangular image, non-indoor scene, inference failure
- [ ] Deployed to dev environment and passes basic load test

---

## References

- LED²-Net paper: *Monocular 360° Layout Estimation via Differentiable Depth Rendering*, Wang et al., CVPR 2021
- LGT-Net paper: *Indoor Panoramic Room Layout Estimation with Geometry-Aware Transformer Network*, Jiang et al., CVPR 2022
- LED²-Net fork: https://github.com/Aroomyaroomy/LED2-Net
- LGT-Net fork: https://github.com/Aroomyaroomy/LGT-Net
